ArmNavigation/seven_joint_arm_to_point_control/PathPlanner.py

```diff
@@ -0,0 +1,398 @@
+        dz = to_node.x[2] - from_node.x[2]
+        d = np.sqrt(np.sum((np.array(to_node.x) - np.array(from_node.x))**2))
+        phi = math.atan2(dy, dx)
+        theta = math.atan2(math.hypot(dx, dy), dz)
+        return d, phi, theta
+
+    @staticmethod
+    def calc_distance_and_angle2(from_node, to_node):
+        dx = to_node.x[0] - from_node.x[0]
+        dy = to_node.x[1] - from_node.x[1]
+        dz = to_node.x[2] - from_node.x[2]
+        d = math.sqrt(dx**2 + dy**2 + dz**2)
+        phi = math.atan2(dy, dx)
+        theta = math.atan2(math.hypot(dx, dy), dz)
+        return d, phi, theta
+
+    @staticmethod
+    def check_collision(node, robot, obstacleList):
+
+        if node is None:
+            return False
+
+        for (ox, oy, oz, size) in obstacleList:
+            for x in node.path_x:
+                x_list, y_list, z_list = robot.get_points(x)
+                dx_list = [ox - x_point for x_point in x_list]
+                dy_list = [oy - y_point for y_point in y_list]
+                dz_list = [oz - z_point for z_point in z_list]
+                d_list = [dx * dx + dy * dy + dz * dz
+                          for (dx, dy, dz) in zip(dx_list,
+                                                  dy_list,
+                                                  dz_list)]
+
+                if min(d_list) <= size ** 2:
+                    return False  # collision
+
+        return True  # safe
+
+
+def main():
+    print("Start " + __file__)
+
+    # init NLinkArm with Denavit-Hartenberg parameters of panda
+    # https://frankaemika.github.io/docs/control_parameters.html
+    # [theta, alpha, a, d]
+    seven_joint_arm = RobotArm([[0., math.pi/2., 0., .333],
+                                [0., -math.pi/2., 0., 0.],
+                                [0., math.pi/2., 0.0825, 0.3160],
+                                [0., -math.pi/2., -0.0825, 0.],
+                                [0., math.pi/2., 0., 0.3840],
+                                [0., math.pi/2., 0.088, 0.],
+                                [0., 0., 0., 0.107]])
+    # ====Search Path with RRT====
+    obstacle_list = [
+        (-.3, -.3, .7, .1),
+        (.0, -.3, .7, .1),
+        (.2, -.1, .3, .15),
+    ]  # [x,y,size(radius)]
+    start = [0 for _ in range(len(seven_joint_arm.link_list))]
+    end = [1.5 for _ in range(len(seven_joint_arm.link_list))]
+    # Set Initial parameters
+    rrt_star = RRTStar(start=start,
+                       goal=end,
+                       rand_area=[0, 2],
+                       max_iter=200,
+                       robot=seven_joint_arm,
+                       obstacle_list=obstacle_list)
+    path = rrt_star.planning(animation=show_animation,
+                             search_until_max_iter=False)
+
+    if path is None:
+        print("Cannot find path")
+    else:
+        print("found path!!")
+
+        # Draw final path
+        if show_animation:
+            ax = rrt_star.draw_graph()
```

Comment:
Can you draw the final target point here.

Reply:
Can you please explain more, do you mean drawing only the final pose?

Reply:
Oh. sorry. I missed this comment.

> do you mean drawing only the final pose?
No.  I mean that it is better to add the target point in the animation. (Maybe, near the tip of the orange arm?)
<img width="456" alt="SS 2020-12-22 at 21 22 24" src="https://user-images.githubusercontent.com/3813847/102888294-d212d000-449b-11eb-9d6a-ed4f0da23c49.png">

Because, I feel it is difficult to know the control is succeeded or not.

Reply:
Aha I see what you mean, but the path planner for an arm is executing in configuration space, in other words the orange arm is the target pose. In fact we fed all joint variables (initial blue arm and final orange arm) to the path planner. This code does not do inverse kinematics, so I do not feed the final tip pose to it to find the path to it, instead, I feed the whole final arm pose in configuration space.
Please let me know if I understood your concern,

Reply:
Oh. OK.  The orange arm is the target configuration. make sense.